tensorflow/python/ops/image_ops_impl.py

```diff
@@ -5581,6 +5581,12 @@ def _sort_scores_and_boxes(scores, boxes):
       x_min, x_max = control_flow_ops.cond(
           x_1_is_min, lambda: (x_1, x_2), lambda: (x_2, x_1))
       boxes = array_ops.concat([y_min, x_min, y_max, x_max], axis=2)
+  else:
+    # TODO(mdanatg/bhack) Without initzializing control_flow_ops.cond variable
+    # in the else branch it is failing autograph control_flow operator
+    # in aug_orelse() verify_tf_cond_branch_ check
+    y_1, x_1, y_2, x_2 = array_ops.split(
+        value=boxes, num_or_size_splits=4, axis=2)
```

Comment:
@mdanatg This is just quick hack to pass the test.
I suppose this emerged as now variables captured in lamba are live_out so _verify_tf_cond_branch_ check in the autograph operator has a constrain check to have these in the else branch.

Reply:
We are not covering this case with the current tests at:

https://github.com/tensorflow/tensorflow/blob/191f0fe3e9de8a7f708a8b0106a1654c797cc573/tensorflow/python/autograph/operators/control_flow_test.py#L1274-L1280

Reply:
It the image_ops failing sample is something like:

```python
    ag__.if_stmt(ag__.not_(ag__.ld(canonicalized_coordinates)), if_body_1, else_body_1, get_state_1, set_state_1, ('boxes', 'x_1', 'x_2', 'y_1', 'y_2'), 5)
```

Reply:
This is the full conversion:
```python
# coding=utf-8
def tf__non_max_suppression_padded_v2(boxes, scores, max_output_size, iou_threshold=None, score_threshold=None, sorted_input=None, canonicalized_coordinates=None, tile_size=None):
 
    with ag__.FunctionScope('non_max_suppression_padded_v2', 'fscope', ag__.ConversionOptions(recursive=True, user_requested=True, optional_features=(), internal_convert_user_code=True)) as fscope:
        do_return = False
        retval_ = ag__.UndefinedReturnValue()

        @ag__.autograph_artifact
        def _sort_scores_and_boxes(scores, boxes):
             with ag__.FunctionScope('_sort_scores_and_boxes', 'fscope_1', ag__.STD) as fscope_1:
                do_return_1 = False
                retval__1 = ag__.UndefinedReturnValue()
                with ag__.ld(ops).name_scope('sort_scores_and_boxes'):
                    batch_size = ag__.converted_call(ag__.ld(array_ops).shape, (ag__.ld(boxes),), None, fscope_1)[0]
                    num_boxes = ag__.converted_call(ag__.ld(array_ops).shape, (ag__.ld(boxes),), None, fscope_1)[1]
                    sorted_scores_indices = ag__.converted_call(ag__.ld(sort_ops).argsort, (ag__.ld(scores),), dict(axis=1, direction='DESCENDING'), fscope_1)
                    index_offsets = ag__.converted_call(ag__.ld(math_ops).range, (ag__.ld(batch_size),), None, fscope_1) * ag__.ld(num_boxes)
                    indices = ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.ld(sorted_scores_indices) + ag__.converted_call(ag__.ld(array_ops).expand_dims, (ag__.ld(index_offsets), 1), None, fscope_1), [-1]), None, fscope_1)
                    sorted_scores = ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.converted_call(ag__.ld(array_ops).gather, (ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.ld(scores), [-1]), None, fscope_1), ag__.ld(indices)), None, fscope_1), [ag__.ld(batch_size), -1]), None, fscope_1)
                    sorted_boxes = ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.converted_call(ag__.ld(array_ops).gather, (ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.ld(boxes), [-1, 4]), None, fscope_1), ag__.ld(indices)), None, fscope_1), [ag__.ld(batch_size), -1, 4]), None, fscope_1)
                try:
                    do_return_1 = True
                    retval__1 = (ag__.ld(sorted_scores), ag__.ld(sorted_boxes), ag__.ld(sorted_scores_indices))
                except:
                    do_return_1 = False
                    raise
                return fscope_1.ret(retval__1, do_return_1)
        batch_dims = ag__.converted_call(ag__.ld(array_ops).shape, (ag__.ld(boxes),), None, fscope)[:-2]
        num_boxes = ag__.converted_call(ag__.ld(array_ops).shape, (ag__.ld(boxes),), None, fscope)[-2]
        boxes = ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.ld(boxes), [-1, ag__.ld(num_boxes), 4]), None, fscope)
        scores = ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.ld(scores), [-1, ag__.ld(num_boxes)]), None, fscope)
        batch_size = ag__.converted_call(ag__.ld(array_ops).shape, (ag__.ld(boxes),), None, fscope)[0]

        def get_state():
            return (boxes, scores)

        def set_state(vars_):
            nonlocal boxes, scores
            (boxes, scores) = vars_

        def if_body():
            nonlocal boxes, scores
            with ag__.ld(ops).name_scope('filter_by_score'):
                score_mask = ag__.converted_call(ag__.ld(math_ops).cast, (ag__.ld(scores) > ag__.ld(score_threshold), ag__.ld(scores).dtype), None, fscope)
                scores = ag__.ld(scores)
                scores *= score_mask
                box_mask = ag__.converted_call(ag__.ld(array_ops).expand_dims, (ag__.converted_call(ag__.ld(math_ops).cast, (ag__.ld(score_mask), ag__.ld(boxes).dtype), None, fscope), 2), None, fscope)
                boxes = ag__.ld(boxes)
                boxes *= box_mask

        def else_body():
            nonlocal boxes, scores
            pass
        score_mask = ag__.Undefined('score_mask')
        box_mask = ag__.Undefined('box_mask')
        ag__.if_stmt(ag__.ld(score_threshold) != ag__.converted_call(ag__.ld(float), ('-inf',), None, fscope), if_body, else_body, get_state, set_state, ('boxes', 'scores'), 2)

        def get_state_1():
            return (boxes, x_1, x_2, y_1, y_2)

        def set_state_1(vars_):
            nonlocal x_1, boxes, y_2, x_2, y_1
            (boxes, x_1, x_2, y_1, y_2) = vars_

        def if_body_1():
            nonlocal x_1, boxes, y_2, x_2, y_1
            with ag__.ld(ops).name_scope('canonicalize_coordinates'):
                (y_1, x_1, y_2, x_2) = ag__.converted_call(ag__.ld(array_ops).split, (), dict(value=ag__.ld(boxes), num_or_size_splits=4, axis=2), fscope)
                y_1_is_min = ag__.converted_call(ag__.ld(math_ops).reduce_all, (ag__.converted_call(ag__.ld(math_ops).less_equal, (ag__.ld(y_1)[0, 0, 0], ag__.ld(y_2)[0, 0, 0]), None, fscope),), None, fscope)
                (y_min, y_max) = ag__.converted_call(ag__.ld(control_flow_ops).cond, (ag__.ld(y_1_is_min), ag__.autograph_artifact(lambda : (ag__.ld(y_1), ag__.ld(y_2))), ag__.autograph_artifact(lambda : (ag__.ld(y_2), ag__.ld(y_1)))), None, fscope)
                x_1_is_min = ag__.converted_call(ag__.ld(math_ops).reduce_all, (ag__.converted_call(ag__.ld(math_ops).less_equal, (ag__.ld(x_1)[0, 0, 0], ag__.ld(x_2)[0, 0, 0]), None, fscope),), None, fscope)
                (x_min, x_max) = ag__.converted_call(ag__.ld(control_flow_ops).cond, (ag__.ld(x_1_is_min), ag__.autograph_artifact(lambda : (ag__.ld(x_1), ag__.ld(x_2))), ag__.autograph_artifact(lambda : (ag__.ld(x_2), ag__.ld(x_1)))), None, fscope)
                boxes = ag__.converted_call(ag__.ld(array_ops).concat, ([ag__.ld(y_min), ag__.ld(x_min), ag__.ld(y_max), ag__.ld(x_max)],), dict(axis=2), fscope)

        def else_body_1():
            nonlocal x_1, boxes, y_2, x_2, y_1
            pass
        x_1 = ag__.Undefined('x_1')
        x_max = ag__.Undefined('x_max')
        x_min = ag__.Undefined('x_min')
        x_2 = ag__.Undefined('x_2')
        y_2 = ag__.Undefined('y_2')
        y_min = ag__.Undefined('y_min')
        y_1_is_min = ag__.Undefined('y_1_is_min')
        x_1_is_min = ag__.Undefined('x_1_is_min')
        y_max = ag__.Undefined('y_max')
        y_1 = ag__.Undefined('y_1')
        ag__.if_stmt(ag__.not_(ag__.ld(canonicalized_coordinates)), if_body_1, else_body_1, get_state_1, set_state_1, ('boxes', 'x_1', 'x_2', 'y_1', 'y_2'), 5)

        def get_state_2():
            return (boxes, scores, sorted_indices)

        def set_state_2(vars_):
            nonlocal boxes, scores, sorted_indices
            (boxes, scores, sorted_indices) = vars_

        def if_body_2():
            nonlocal boxes, scores, sorted_indices
            (scores, boxes, sorted_indices) = ag__.converted_call(ag__.ld(_sort_scores_and_boxes), (ag__.ld(scores), ag__.ld(boxes)), None, fscope)

        def else_body_2():
            nonlocal boxes, scores, sorted_indices
            sorted_indices = ag__.converted_call(ag__.ld(array_ops).zeros_like, (ag__.ld(scores),), dict(dtype=ag__.ld(dtypes).int32), fscope)
        sorted_indices = ag__.Undefined('sorted_indices')
        ag__.if_stmt(ag__.not_(ag__.ld(sorted_input)), if_body_2, else_body_2, get_state_2, set_state_2, ('boxes', 'scores', 'sorted_indices'), 3)
        pad = ag__.converted_call(ag__.ld(math_ops).cast, (ag__.converted_call(ag__.ld(math_ops).ceil, (ag__.converted_call(ag__.ld(math_ops).cast, (ag__.converted_call(ag__.ld(math_ops).maximum, (ag__.ld(num_boxes), ag__.ld(max_output_size)), None, fscope), ag__.ld(dtypes).float32), None, fscope) / ag__.converted_call(ag__.ld(math_ops).cast, (ag__.ld(tile_size), ag__.ld(dtypes).float32), None, fscope),), None, fscope), ag__.ld(dtypes).int32), None, fscope) * ag__.ld(tile_size) - ag__.ld(num_boxes)
        boxes = ag__.converted_call(ag__.ld(array_ops).pad, (ag__.converted_call(ag__.ld(math_ops).cast, (ag__.ld(boxes), ag__.ld(dtypes).float32), None, fscope), [[0, 0], [0, ag__.ld(pad)], [0, 0]]), None, fscope)
        scores = ag__.converted_call(ag__.ld(array_ops).pad, (ag__.converted_call(ag__.ld(math_ops).cast, (ag__.ld(scores), ag__.ld(dtypes).float32), None, fscope), [[0, 0], [0, ag__.ld(pad)]]), None, fscope)
        num_boxes_after_padding = ag__.ld(num_boxes) + ag__.ld(pad)
        num_iterations = ag__.ld(num_boxes_after_padding) // ag__.ld(tile_size)

        @ag__.autograph_artifact
        def _loop_cond(unused_boxes, unused_threshold, output_size, idx):
            with ag__.FunctionScope('_loop_cond', 'fscope_2', ag__.STD) as fscope_2:
                do_return_2 = False
                retval__2 = ag__.UndefinedReturnValue()
                try:
                    do_return_2 = True
                    retval__2 = ag__.converted_call(ag__.ld(math_ops).logical_and, (ag__.converted_call(ag__.ld(math_ops).reduce_min, (ag__.ld(output_size),), None, fscope_2) < ag__.ld(max_output_size), ag__.ld(idx) < ag__.ld(num_iterations)), None, fscope_2)
                except:
                    do_return_2 = False
                    raise
                return fscope_2.ret(retval__2, do_return_2)

        @ag__.autograph_artifact
        def suppression_loop_body(boxes, iou_threshold, output_size, idx):
            with ag__.FunctionScope('suppression_loop_body', 'fscope_3', ag__.STD) as fscope_3:
                do_return_3 = False
                retval__3 = ag__.UndefinedReturnValue()
                try:
                    do_return_3 = True
                    retval__3 = ag__.converted_call(ag__.ld(_suppression_loop_body), (ag__.ld(boxes), ag__.ld(iou_threshold), ag__.ld(output_size), ag__.ld(idx), ag__.ld(tile_size)), None, fscope_3)
                except:
                    do_return_3 = False
                    raise
                return fscope_3.ret(retval__3, do_return_3)
        (selected_boxes, _, output_size, _) = ag__.converted_call(ag__.ld(control_flow_ops).while_loop, (ag__.ld(_loop_cond), ag__.ld(suppression_loop_body), [ag__.ld(boxes), ag__.ld(iou_threshold), ag__.converted_call(ag__.ld(array_ops).zeros, ([ag__.ld(batch_size)], ag__.ld(dtypes).int32), None, fscope), ag__.converted_call(ag__.ld(constant_op).constant, (0,), None, fscope)]), dict(shape_invariants=[ag__.converted_call(ag__.ld(tensor_shape).TensorShape, ([None, None, 4],), None, fscope), ag__.converted_call(ag__.ld(tensor_shape).TensorShape, ([],), None, fscope), ag__.converted_call(ag__.ld(tensor_shape).TensorShape, ([None],), None, fscope), ag__.converted_call(ag__.ld(tensor_shape).TensorShape, ([],), None, fscope)]), fscope)
        num_valid = ag__.converted_call(ag__.ld(math_ops).minimum, (ag__.ld(output_size), ag__.ld(max_output_size)), None, fscope)
        idx = ag__.ld(num_boxes_after_padding) - ag__.converted_call(ag__.ld(math_ops).cast, (ag__.converted_call(ag__.ld(nn_ops).top_k, (ag__.converted_call(ag__.ld(math_ops).cast, (ag__.converted_call(ag__.ld(math_ops).reduce_any, (ag__.ld(selected_boxes) > 0, [2]), None, fscope), ag__.ld(dtypes).int32), None, fscope) * ag__.converted_call(ag__.ld(array_ops).expand_dims, (ag__.converted_call(ag__.ld(math_ops).range, (ag__.ld(num_boxes_after_padding), 0, -1), None, fscope), 0), None, fscope), ag__.ld(max_output_size)), None, fscope)[0], ag__.ld(dtypes).int32), None, fscope)
        idx = ag__.converted_call(ag__.ld(math_ops).minimum, (ag__.ld(idx), ag__.ld(num_boxes) - 1), None, fscope)

        def get_state_3():
            return (idx,)

        def set_state_3(vars_):
            nonlocal idx
            (idx,) = vars_

        def if_body_3():
            nonlocal idx
            index_offsets = ag__.converted_call(ag__.ld(math_ops).range, (ag__.ld(batch_size),), None, fscope) * ag__.ld(num_boxes)
            gather_idx = ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.ld(idx) + ag__.converted_call(ag__.ld(array_ops).expand_dims, (ag__.ld(index_offsets), 1), None, fscope), [-1]), None, fscope)
            idx = ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.converted_call(ag__.ld(array_ops).gather, (ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.ld(sorted_indices), [-1]), None, fscope), ag__.ld(gather_idx)), None, fscope), [ag__.ld(batch_size), -1]), None, fscope)

        def else_body_3():
            nonlocal idx
            pass
        index_offsets = ag__.Undefined('index_offsets')
        gather_idx = ag__.Undefined('gather_idx')
        ag__.if_stmt(ag__.not_(ag__.ld(sorted_input)), if_body_3, else_body_3, get_state_3, set_state_3, ('idx',), 1)
        invalid_index = ag__.converted_call(ag__.ld(array_ops).zeros, ([ag__.ld(batch_size), ag__.ld(max_output_size)],), dict(dtype=ag__.ld(dtypes).int32), fscope)
        idx_index = ag__.converted_call(ag__.ld(array_ops).expand_dims, (ag__.converted_call(ag__.ld(math_ops).range, (ag__.ld(max_output_size),), None, fscope), 0), None, fscope)
        num_valid_expanded = ag__.converted_call(ag__.ld(array_ops).expand_dims, (ag__.ld(num_valid), 1), None, fscope)
        idx = ag__.converted_call(ag__.ld(array_ops).where, (ag__.ld(idx_index) < ag__.ld(num_valid_expanded), ag__.ld(idx), ag__.ld(invalid_index)), None, fscope)
        num_valid = ag__.converted_call(ag__.ld(array_ops).reshape, (ag__.ld(num_valid), ag__.ld(batch_dims)), None, fscope)
        try:
            do_return = True
            retval_ = (ag__.ld(idx), ag__.ld(num_valid))
        except:
            do_return = False
            raise
        return fscope.ret(retval_, do_return)
```

Reply:
So if you see the problem is exactly here this is the `else_body` (augmented) without our else dummy implementation:
```python 
if not canonicalized_coordinates: 
```

```python

        def else_body_1():
            nonlocal x_1, boxes, y_2, x_2, y_1
            pass
```

https://github.com/tensorflow/tensorflow/blob/7bcaba629b7d5022b654ba6ed78ad7470ce412aa/tensorflow/python/ops/image_ops_impl.py#L5571-L5589

Reply:
Implementing our else workarund is:
```python
def else_body_1():
            nonlocal x_1, boxes, y_2, y_1, x_2
            (y_1, x_1, y_2, x_2) = ag__.converted_call(ag__.ld(array_ops).split, (), dict(value=ag__.ld(boxes), num_or_size_splits=4, axis=2), fscope)
```

And it pass the check.

Reply:
Basically it is the same case of few line below in master `sorted_indices` (nonlocal and undefined like are our lambda args):

https://github.com/tensorflow/tensorflow/blob/7806ae568aba65ed87fccec95d72151f0365939c/tensorflow/python/ops/image_ops_impl.py#L5587-L5591

Reply:
Very nice analysis! I agree with your conclusions, but I fear that even though correct, we would cause quite a bit of disruption for many users; most might be unable to fix it quickly. 

Even with safeguards in place, we might still break enough people to force a rollback, but I think that's worth giving a shot.

So what I'm thinking is adding an AST annotation to all the lambdas we create internally, to denote them to be "used locally", and that static analysis can assume they will not be used later.

Reply:
Just a reminder. We decided that we need to talk about how to track in source lambdas when they stored in non-local containers (for a late evaluation).

Reply:
A small example:
```python
import tensorflow as tf
@tf.function
def test(b):
  if (b):
    x=1
    y=2
    z=3
    p = tf.cond([True], lambda: x, lambda: y)
  else:
    p = 0
    x = 0
    y = 0
  return(p)
print(test(tf.constant(True)))
```

We cannot explicit remove the else branch or remove `p, x, y` initializations in the else branch as these are mandatory. Of course we could exclude `z` init in the else branch:

 ```python
ValueError: 'p'  must also be initialized in the else branch
````